### PR TITLE
Introduce Thread-local storage variable to reduce atomic contention when update used_memory metrics.

### LINKED ISSUE #467: Significance of absolute accurate `used_memory`?
**The problem/use-case that the feature addresses**

Currently, when call the zmalloc/zfree related functions, it will call the zmalloc_size and atomic operations to update `used_memory`, these are all costly operations, is it worth to put such expensive operation in a frequently low API?

Some pull requests are submitted to optimize for such kind of this issue.

https://github.com/valkey-io/valkey/pull/308
https://github.com/valkey-io/valkey/pull/453 

**Alternatives you've considered**

Maybe we can consider to remove the absolute accurate property of `used_memory` to trade off between the performance?

### PR DESCRIPTION
**Pull request by zhulipeng**
#### Description 
This patch try to introduce Thread-local storage variable to replace atomic for `zmalloc` to reduce unnecessary contention. 

#### Problem Statement
 `zmalloc` and `zfree` related functions will update the `used_memory` for each operation, and they are called very frequency.  From the benchmark of [memtier_benchmark-1Mkeys-load-stream-5-fields-with-100B-values-pipeline-10.yml](https://github.com/redis/redis-benchmarks-specification/blob/main/redis_benchmarks_specification/test-suites/memtier_benchmark-1Mkeys-load-stream-5-fields-with-100B-values-pipeline-10.yml) , the cycles ratio of `zmalloc` and `zfree` are high, they are wrappers for the lower allocator library, it should not take too much cycles. And most of the cycles are contributed by `lock add` and `lock sub` , they are expensive instructions. From the profiling, the metrics' update mainly come from the main thread, use a TLS will reduce a lot of contention. 

image

image

##### Test Env

- OS: CentOS Stream 8
- Kernel: 6.2.0
- Platform: Intel Xeon Platinum 8380
- Server and Client in same socket

##### Start Server 
`taskset -c 0 ~/valkey/src/valkey-server /tmp/valkey_1.conf`
```
port 9001
bind * -::*
daemonize yes
protected-mode no
save ""
```

By using the benchmark [memtier_benchmark-1Mkeys-load-stream-5-fields-with-100B-values-pipeline-10.yml](https://github.com/redis/redis-benchmarks-specification/blob/main/redis_benchmarks_specification/test-suites/memtier_benchmark-1Mkeys-load-stream-5-fields-with-100B-values-pipeline-10.yml).
`memtier_benchmark -s 127.0.0.1 -p 9001  "--pipeline" "10" "--data-size" "100" --command "XADD __key__ MAXLEN ~ 1 * field __data__" --command-key-pattern="P" --key-minimum=1 --key-maximum 1000000 --test-time 180 -c 50 -t 4 --hide-histogram`

We can observe more than **6%** QPS gain.  

For the other benchmarks SET/GET, using the commands like:
`taskset -c 6-9  ~/valkey/src/valkey-benchmark -p 9001 -t set,get -d 100 -r 1000000 -n 1000000 -c 50 --threads 4`

No perf gain and regression.

With pipeline enabled,  I can observe 4% perf gain with [test case](https://github.com/redis/redis-benchmarks-specification/blob/main/redis_benchmarks_specification/test-suites/memtier_benchmark-1Mkeys-load-string-with-100B-values-pipeline-10.yml).
`taskset -c 4-7 memtier_benchmark -s 127.0.0.1 -p 9001 "--pipeline" "10" "--data-size" "100" --ratio 1:0 --key-pattern P:P --key-minimum=1 --key-maximum 1000000 --test-time 180 -c 50 -t 4 --hide-histogram`

### CODE REVIEW

File: src/zmalloc.c
Comment (madolson) [inline src/zmalloc.c]:
```suggestion
#define update_zmalloc_stat_alloc(__n) used_memory_thread[thread_index] += (__n)
#define update_zmalloc_stat_free(__n) used_memory_thread[thread_index] -= (__n)
```
One of my greatest pet peeves is naming thread local storage TLS, because my mind only thinks it means transport layer security :)

File: src/zmalloc.c
Comment (madolson) [inline src/zmalloc.c]:
```suggestion
#define MAX_THREADS_NUM (IO_THREADS_MAX_NUM + BIO_THREAD_COUNT + 1)
```

Can we do something like this? I think we need to implement a new define for BIO thread count. Then we don't have to worry about this changing.

File: src/zmalloc.c
Comment (madolson) [inline src/zmalloc.c]:
```suggestion
        serverAssert(i < MAX_THREADS_NUM);
        um += used_memory_tls[i];
```

File: src/zmalloc.c
Comment (madolson) [inline src/zmalloc.c]:
```suggestion
    serverAssert(total_active_threads < MAX_THREADS_NUM);
    atomicGetIncr(total_active_threads, thread_index, 1);
```

File: src/zmalloc.c
Comment (madolson) [inline src/zmalloc.c]:
```suggestion
static unsigned long long used_memory_tls[MAX_THREADS_NUM];
```
The BIO threads will underflow, since it's only releasing memory, and the main thread will likely overflow if there is a BIO thread freeing memory. Although I don't think this is technically an issue, I think it would be clearer to indicate that with a signed variable then leaving it unsigned.

**Review by madolson — COMMENTED** [review]
What are the server configurations you are using for the test? I didn't see them listed. Can you also just do a simple test with get/set?

**Comment by zhulipeng** [comment]
> What are the server configurations you are using for the test? I didn't see them listed. Can you also just do a simple test with get/set?

Just update the server config and SET/GET results in top comment.

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c]: `IO_THREADS_MAX_NUM` is defined in `networking.c` and `BIO_THREAD_COUNT` has a similar define `BIO_WORKER_NUM` in `bio.c`. How do we reuse the define in the `zmalloc.c` ? It is a very low level API.

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c]: Thanks for the suggestion, changed.

File: src/zmalloc.c
Comment (qlong) [inline src/zmalloc.c]: Wondering how important is it for zmalloc_used_memory to return the correct used_memory? As we loop through  to add up the memory used by each thread, the threads that had been counted may continue to allocate or to free memory, so the sum might under or over account the actual memory used by all threads. 

I don't think this is an issue if the function is only used for metrics, but overMaxmemoryAfterAlloc (evict.c) is not for metrics,  we probably should ensure (and call out) that undercounting memory usage would not be an issue there

File: src/zmalloc.c
Comment (qlong) [inline src/zmalloc.c]: thread_index is thread local, but used_memory_thread[thread_index] is not.  Without some sort of locking, a thread might get strange value from used_memory_thread[thread_index]  if another thread is updating it

File: src/zmalloc.c
Comment (qlong) [inline src/zmalloc.c]: I think used_memory_thread[i] can return corrupted data if thread i is updating the value at the same time

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c]: @qlong I understand your concern, at the beginning, I want to keep an approximate value for used_memory, each thread keep its own used_memory_thread, only used_memory_thread greater than a threshold(e.g. 1k) then update to global atomic used_memory, the pseudocode will be like this, but I saw the test case in https://github.com/valkey-io/valkey/blob/unstable/src/zmalloc.c#L917, seems it need an accurate used_memory? Back to this PR, each element in used_memory_thread array only have one writer to update the value, when read, the reader may have gap between `oldvalue` and `newvalue`, depends on the read timing. 
@madolson  What's your thought?
```diff
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define update_zmalloc_stat_alloc(__n) do { \
+    used_memory_thread += (__n); \
+    if (used_memory_thread > 1024) atomicIncr(used_memory, used_memory_thread); \
+} while (0) \
+
+#define update_zmalloc_stat_free(__n) do { \
+    used_memory_thread -= (__n); \
+    if (used_memory_thread > 1024) atomicDecr(used_memory, used_memory_thread); \
+} while (0) \

 static serverAtomic size_t used_memory = 0;
+static __thread size_t used_memory_thread = 0;
```

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c]: What do you mean about the `strange value`? The reader can only read `old_value` or `new_value` in this case.

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c]: Ditto.

**Comment by zhulipeng** [comment]
@valkey-io/core-team, could you help to take a look at this patch?

File: src/zmalloc.c
Comment (qlong) [reply src/zmalloc.c]: @lipzhu I meant corrupted value due to race condition among threads. I did not see (maybe I have missed) how we guarantee reader read either old or new value. The reader can see corrupted value of used_memory_thread[i] if the writer is updating used_memory_thread[i] at the same time

Comment (zhulipeng) [reply src/zmalloc.c]: Yes, the semantics is a little different from the origin implementation, but do we really need such accurate `used_memory`? The cost is to add the expensive `lock add/sub` instruction in the low-level API like zmalloc/zfree functions, especially most calls to these functions come from the main thread. 
And the `new_value` will be seen in next read.

Comment (qlong) [reply src/zmalloc.c]: @lipzhu I think data corruption caused by race condition is not about accuracy, the corrupted data can be random from 0 to 2^64-1. This can lead to system correctness or stability issue. 

The original implementation uses atomic number, which leverages support from CPU and is quite efficient as counter that gets updated frequently from different threads.

**Review by enjoy-binbin — COMMENTED** [review]
i did not take a deep look, how do we handle module threads?

**Comment by zhulipeng** [comment]
Hi @enjoy-binbin , thanks for your comments for this patch. BTW, your blog in yuque help me a lot in understanding redis/valkey. 

> i did not take a deep look, how do we handle module threads?

Sorry I missed the modules thread, maybe I can remove the explicit call of the `zmalloc_register_thread_index` in start routine, initialize  `static __thread int thread_index = -1` and then add a checker `if (thread_index == -1) zmalloc_register_thread_index() `in zmalloc/zfree?

**Comment by zhulipeng** [comment]
@enjoy-binbin @PingXie @madolson Thoughts about this patch?

File: src/zmalloc.c
Comment (PingXie) [reply src/zmalloc.c]: yeah TLS is unfortunatelly too popular :). `_thread` works for me too.

File: src/zmalloc.c
Comment (PingXie) [reply src/zmalloc.c]: Yeah there is a layering concern here. Can you add a `c_assert` to make sure the condition of `(IO_THREADS_MAX_NUM + BIO_THREAD_COUNT + 1) < IO_THREADS_MAX_NUM` is not violated?

File: src/zmalloc.c
Comment (PingXie) [reply src/zmalloc.c]: Since both `total_active_threads` and `thread_index` are of `size_t`, whose sizes match the native word size of the platform, the proposed change should work. 

Strictly speaking though (if `size_t` was 64 bit on a 32-bit platform), L112 and L113 need to be flipped. We also shouldn't touch the global var `total_active_threads` without the atomic operation; instead, we should access the thread local var `thread_index`. 

```suggestion
    atomicGetIncr(total_active_threads, thread_index, 1);
    serverAssert(thread_index < MAX_THREADS_NUM);
```

File: src/zmalloc.c
Comment (PingXie) [reply src/zmalloc.c]: btw, if we go with this route, we should probably introduce a new header to host all thread local storage related definition. It feels a bit out of place to define `MAX_THREADS_NUM` in zmalloc.c. We might have other consumers of this macro in the future. I wonder if we should call the new header `system.h`, which would be at the lowest layer.

File: src/zmalloc.c
Comment (PingXie) [reply src/zmalloc.c]: a few quick thoughts

1. I see value in removing atomic operations
2. Since we use `size_t` here for the counters, which should be natively aligned as well, the worst we can get would be staleness as opposed to inconsistency.
3. The absolute accuracy doesn't have much value for metrics. The value will get stale immediately after the `INFO` command returns, whether it is absolutely accurate or not. As long as the value is close enough I think we are good. We can change the tests
4. Agreed that `overMaxmemoryAfterAlloc` is a higher bar than metrics. My gut feeling is that the staleness brought by the relaxed memory access should still be negligible. I doubt we could accumulate a lot of changes before the CPU flushes out its cache to the main memory. Curious to know if anyone has more definitive information.
5. The absolute accuracy requirement is directly at odds with the performance overhead incurred by the atomic operation. So one of them has to go.

Comment (zhulipeng) [reply src/zmalloc.c]: Seems we all had concern about the absolute accuracy of `used_memory` in `overMaxmemoryAfterAlloc`, IMO, this is not a problem, because in the definition of `overMaxmemoryAfterAlloc` , the comparison between `moremem` and `used_memory` (Their semantics are different) is not absolute accuracy if we consider the gap between requested memory and actual allocated memory from allocator.  
``` c
int overMaxmemoryAfterAlloc(size_t moremem) {
    if (!server.maxmemory) return  0; /* No limit. */

    /* Check quickly. */
    size_t mem_used = zmalloc_used_memory();
    if (mem_used + moremem <= server.maxmemory) return 0;

    size_t overhead = freeMemoryGetNotCountedMemory();
    mem_used = (mem_used > overhead) ? mem_used - overhead : 0;
    return mem_used + moremem > server.maxmemory;
}
```

Comment (zhulipeng) [reply src/zmalloc.c]: Kindly ping @PingXie , any comments about this, do we need to continue this?

Comment (PingXie) [reply src/zmalloc.c]: I think there is a way to address this concern while retaining the most of the performance benefit if we keep the existing io-threading model. The idea would be to keep a mirrored per-thread counter array such that each thread periodically "flushes" its thread local counter value to the corresponding element in the global per-thread counter array using atomic operations. This would work with the current io-threading model where io-threads and main threads run in lock-steps. Every io thread would need to push its local value to this global array when it is done with its work, but only for once. Then by the time the main thread starts executing, it would have the accurate data. We can reduce the number of atomic operations to one per io-thread per epoll session. 

We would have to revisit this if we went for a different threading model (such as #22) .

That said, I am still not sure how accurate this information needs to be.  `zmalloc_used_memory` doesn't represent the RSS usage of the process anyways. I think it would be really helpful if we could somehow establish the worst possible case, empirically or even a SWAG.

Comment (zhulipeng) [reply src/zmalloc.c]: > That said, I am still not sure how accurate this information needs to be. `zmalloc_used_memory` doesn't represent the RSS usage of the process anyways.
>

IMO, there is no meaningful of the absolute accurate `zmalloc_used_memory`, to achieve the goal, the cost is very high, not only the expensive atomic operation in every `zmalloc/zfree` operation, but also the call of `je_malloc_usable_size` is very high (**From the cycles hotspot snapshot in top comment, you can find the cycles ratio of `je_malloc_usable_size` is the top 1 hotspot**). If we can accept the a little latency of `zmalloc_used_memory`,  ideally we may have at least **10%** perf gain. 

@PingXie Do you think we need open an issue to discuss with the rest of @valkey-io/core-team of the necessary of absolute accurate `zmalloc_used_memory` ?

Comment (PingXie) [reply src/zmalloc.c]: I saw you opened #467. Thanks @lipzhu.

**Comment by zhulipeng** [comment]
Kindly ping @valkey-io/core-team.

File: src/zmalloc.c
Comment (zuiderkwast) [inline src/zmalloc.c]: Minor: We could use a static inline function for this instead of a macro.

**Review by zuiderkwast — COMMENTED** [review]
This one seems almost ready to merge.

@PingXie @enjoy-binbin @madolson are there any open questions and suggestions how we can close them?

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c]: Done.

**Comment by zuiderkwast** [comment]
In the Valkey Contributor Summit, someone (who?) suggested that we remove this counter in zmalloc and instead rely on the stats from the allocator, like `je_mallctl("stats.allocated")`. I tend to agree with that. zmalloc is in a hot code path and it'd be good to avoid tracking this. The only difference seems to be that we can see how much was allocated using zmalloc vs allocated in other ways (by libs, etc.). Is there any good reason to see the difference between these? 

The INFO fields are `used_memory` (from zmalloc counter) vs `allocator_allocated` (stats from allocator).  I imagine we can simply let `used_memory` be identical to `allocator_allocated`.

**Comment by PingXie** [comment]
I think there are still 2 comments to be addressed

1. the hard-coded max thread number
2. `zmalloc_used_memory` can be even more inaccurate now with the PR. I kind of like @lipzhu's proposal at https://github.com/valkey-io/valkey/pull/308#discussion_r1566641483 but I think there is a bug in the code snippet - shouldn't `used_memory_thread` be cleared after getting applied to `used_memory`? In other words, I think `used_memory_thread` should be used to track/cache the deltas that haven't been applied to the global counter.

>je_mallctl("stats.allocated")

@lipzhu you mentioned that `je_malloc_usable_size` is very expensive. I am wondering where this call would stand?

**Comment by zhulipeng** [comment]
> I imagine we can simply let used_memory be identical to allocator_allocated.

@zuiderkwast I am not clear the full history of `used_memory`, but I saw comments like below, the `used_memory` is designed to be different with `allocator_allocated`? And seems we only have `allocator_allocated` with jemalloc now. 

```
    /* Unlike zmalloc_used_memory, this matches the stats.resident by taking
     * into account all allocations done by this process (not only zmalloc). */
    je_mallctl("stats.allocated", allocated, &sz, NULL, 0);
```
But I noticed that the metrics of `used_memory` and `allocator_allocated` are very closed during the benchmark test. 
```
info memory
used_memory:46356571552
allocator_allocated:46356604048
```


> the hard-coded max thread number

@PingXie Yes, I am also a little confused here, according to the suggestion https://github.com/valkey-io/valkey/pull/308#discussion_r1564891132, but how do we handle the modules, is there any way to get the modules number to be a factor of max thread number?

> zmalloc_used_memory can be even more inaccurate now with the PR. 

Can you be more specific.  

>I kind of like @lipzhu's proposal at https://github.com/valkey-io/valkey/pull/308#discussion_r1566641483 but I think there is a bug in the code snippet - shouldn't used_memory_thread be cleared after getting applied to used_memory? In other words, I think used_memory_thread should be used to track/cache the deltas that haven't been applied to the global counter.

I am OK with any proposal if your core-team make the final decision. 

> @lipzhu you mentioned that je_malloc_usable_size is very expensive. I am wondering where this call would stand?

Each call of `zmalloc` will also call `zmalloc_size` to get the allocated size. And I observed the `je_malloc_usable_size` take about 6% CPU cycles.

**Comment by zuiderkwast** [comment]
Thanks for putting some focus on this performance bottleneck!

I commented in the issue #467.

I think we can use jemalloc's stats and fallback to counting in zmalloc only if another allocator is used.

**Comment by PingXie** [comment]
>Each call of zmalloc will also call zmalloc_size to get the allocated size. And I observed the je_malloc_usable_size take about 6% CPU cycles.

Sorry I meant to ask if you had a chance to evaluate the overhead of `je_mallctl("stats.allocated")`

> I think we can use jemalloc's stats and fallback to counting in zmalloc only if another allocator is used.

I think we should get a perf reading of this change first. 

>Thanks for putting some focus on this performance bottleneck!

+1. Appreciate it, @lipzhu!

**Comment by zhulipeng** [comment]
@zuiderkwast @PingXie Update the patch as suggested in https://github.com/valkey-io/valkey/issues/467.

File: src/zmalloc.c
Comment (zuiderkwast) [inline src/zmalloc.c]: `__thread` is a compiler extension. We can use standard C11 `thread_local` if we include `<threads.h>`. See https://en.cppreference.com/w/c/language/storage_duration#Notes. (`_Thread_local` is another possibility C11 without include threads.h.)

File: src/zmalloc.c
Comment (zuiderkwast) [inline src/zmalloc.c:415]: How can it become negative? Only if we allocate in one thread and free in another?

**Review by zuiderkwast — APPROVED** [review]
LGTM (with some nit)

It's almost the code I posted in a comment. :) It was more fun to write code than to explain in text. Sorry for that.

File: src/zmalloc.c
Comment (zhulipeng) [reply src/zmalloc.c:415]: Yes, allocate in main thread and free in bio thread?

**Comment by zuiderkwast** [comment]
It seems we have a problem on macos:

```
zmalloc.c:55:10: fatal error: 'threads.h' file not found
#include <threads.h>
         ^~~~~~~~~~~
```

See https://stackoverflow.com/questions/16244153/clang-c11-threads-h-not-found

Maybe we need some conditional compilation like

```c
#if __STDC_NO_THREADS__
#define thread_local __thread
#else
#include <threads.h>
#endif
```

**Review by zuiderkwast — APPROVED** [review]
LGTM.

There's a "major-decision-pending" label on the issue, so let's wait for a majority of the @valkey-io/core-team (vote or approve) before we merge this.

**Comment by soloestoy** [comment]
I have some concerns:
1. Now the threshold is 1MB, if this threshold is not reached, `used_memory` would not be updated. Some special cases, for example lazyfree, bio only frees 900KB memory, then no new jobs comes, `used_memory` will be inaccurate forever. I think we need a mechanism to compensate in order to achieve eventual consistency, not only through size thresholds but also time thresholds, such as updating the delta every 1 second.
2. We have many threads, main thread, bio, io-threads, module threads. Too many threads may lead to larger errors, as in the situation in point 1, the error is the number of threads times 1MB. If the problem in point 1 can be resolved, then this would not be an issue either.

**Comment by zuiderkwast** [comment]
> 1. also time thresholds, such as updating the delta every 1 second

@soloestoy Isn't checking time in the allocator function too slow? It is a very hot code path.

Some alternative ideas:

* Use 100KB or 10KB instead of 1M to make the error smaller? Most of the allocations are small like less than 1K anyway.
* Add some explicit call to flush this delta to the atomic variable, such as after each job in bio.c and IO threads.

**Comment by soloestoy** [comment]
>Isn't checking time in the allocator function too slow? It is a very hot code path.

yes, it would be best if every thread can register a timer to trigger updating delta every second.

>Use 100KB or 10KB instead of 1M to make the error smaller? Most of the allocations are small like less than 1K anyway.

maybe 10KB is an acceptable error.

>Add some explicit call to flush this delta to the atomic variable, such as after each job in bio.c and IO threads.

native threads can work well, but module's thread is a problem.

**Comment by zuiderkwast** [comment]
> it would be best if every thread can register a timer to trigger updating delta every second.

@soloestoy How to do this per thread in a simple way? Signal handler?

> maybe 10KB is an acceptable error.

Let's do this then? I think it can provide much of the benefit already.

> native threads can work well, but module's thread is a problem.

Simple solution: We can add do explicit flush to the atomic counter in ValkeyModule_Alloc and ValkeyModule_Free.

**Comment by zhulipeng** [comment]
@soloestoy The PR is based on the assumption that we can accept `used_memory` is not accurate, the worst gap is just as you mentioned threads_num *  1M, you can refer @PingXie and @zuiderkwast discussion https://github.com/valkey-io/valkey/issues/467 for more details. 

If we are not alignment of the absolute `used_memory`, I want to recall another proposal (actually the initial version of this PR), use a thread-local storage to maintain each thread's own `used_memory` and flush to `used_memory` through `zmalloc_used_memory()`, it is a bit like the opposite of rwlock, write parallelly and read exclusively. I create another 
  https://github.com/valkey-io/valkey/pull/674 to demonstrate the proposal for you core members to review and compare.

**Pull request closed without merging**
